rfd/0143-external-k8s-joining.md

```diff
@@ -0,0 +1,319 @@
+
+message ProvisionTokenSpecV2KubernetesStaticJWKS{
+  // JWKS contains the JSON Web Keys to validate a service account token
+  // against. This should be copied from the Kubernetes server's JWKS endpoint
+  // (/openid/v1/jwks).
+  string JWKS = 1;
+}
+
+message ProvisionTokenSpecV2Kubernetes{
+  // Rule is a set of properties the Kubernetes-issued token might have to be
+  // allowed to use this ProvisionToken
+  message Rule {
+    // ServiceAccount is the namespaced name of the Kubernetes service account.
+    // Its format is "namespace:service-account".
+    string ServiceAccount = 1;
+  }
+  
+  // Allow is a list of Rules, nodes using this token must match one
+  // allow rule to use this token.
+  repeated Rule Allow = 1;
+
+  // Type specifies what behaviour should be used for validating a
+  // Kubernetes Service Account token. This must be
+  // - `in_cluster`
+  // - `static_jwks`
+  // If omitted, it will default to `in_cluster`.
+  string Type = 2;
+
+  // StaticJWKS contains configuration specific to the `static_jwks` type.
+  ProvisionTokenSpecV2KubernetesStaticJWKS StaticJWKS = 3;
+}
+```
+
+In order to obtain the `jwks` value for their Kubernetes Cluster, operators
+will need to run the following steps from a machine already authenticated
+with their cluster:
+
+- `kubectl proxy -p 8080`
+- `curl http://localhost:8080/openid/v1/jwks`
+
+In the case that an operator rotates the certificate authority used by their
+Kubernetes Cluster to sign Service Account JWTS, the `jwks` field will also
+need to be updated.
+
+#### Configuring Kubernetes
+
+The joining entity will need to exist in a pod where a Service Account token
+has been projected with an audience equal to the name of the Teleport cluster.
+
+E.g for a Teleport cluster named: "example.teleport.sh":
+
+```yaml
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: my-service-account
+---
+apiVersion: v1
+kind: Pod
+metadata:
+  name: nginx
+spec:
+  containers:
+    - image: nginx
+      name: nginx
+      volumeMounts:
+        - mountPath: /var/run/secrets/tokens
+          name: my-service-account
+  serviceAccountName: my-service-account
+  volumes:
+    - name: my-service-account
+      projected:
+        sources:
+          - serviceAccountToken:
+              path: my-service-account
+              expirationSeconds: 600
+              audience: example.teleport.sh
+```
+
+### Implementation
+
+The `static_jwks` variant will leverage some of the same parts of the flow
+as the `in_cluster` variant.
+
+1. As with `in_cluster`, the client loads the service account JWT from the
+  filesystem that was mounted by Kubelet and calls `RegisterUsingToken` with 
+  this JWT and the name of the token it wishes to join using.
+2. The Auth Server finds the ProvisionToken matching the name of token. If this
+  is a ProvisionToken with `type: static_jwks`, the flow now diverges from the
+  existing `in_cluster` implementation.
+5. The Auth Server validates:
```

Comment:
Does the JWT also include an `issuer` field? If so might be worth to check it explicetely too.

Reply:
https://github.com/gravitational/teleport/pull/29225/commits/cffd6b62f46513844da49fede3a58161c21db327